Automattic/jetpack
# autoloader: Add debug checks for common confusing situations

## Proposed changes:

A common cause of confusing errors is when a plugin autoloads classes during the plugin load, before the 'plugins_loaded' hook. If that plugin has an older version of the class, that older version may be loaded before a plugin providing the newer version of the class has a chance to register. Even more confusingly, this will likely be intermittent, only showing up when the autoloader's plugin cache is invalidated. To debug this situation, you can set the `JETPACK_AUTOLOAD_DEBUG_EARLY_LOADS` constant to true.

Another common cause of confusing errors is when a plugin registers its own autoloader at a higher priority than the Jetpack Autoloader, and that autoloader would load packages that should be handled by the Jetpack Autoloader. Setting the `JETPACK_AUTOLOAD_DEBUG_CONFLICTING_LOADERS` constant to true will check for standard Composer autoloaders with such a conflict.

### Other information:


## Jetpack product discussion


p1724248864310869/1724237275.205199-slack-C05Q5HSS013

## Does this pull request change what data or activity we track or use?


No

## Testing instructions:






* Set the constants, try it out.
* For `JETPACK_AUTOLOAD_DEBUG_CONFLICTING_LOADERS` in particular, installing the woocommerce-shipping plugin v1.0.4 should have it report the problem.
